doc/cprover-manual/loop-invariant-synthesizer.md

```diff
@@ -0,0 +1,28 @@
+[CPROVER Manual TOC](../)
+
+## Loop Invariant Synthesis with goto-instrument
+
+This is a loop invariant synthesizer built in goto-instrument that 
+synthesize [invariant clause](contracts-invariants.md) for a 
+pre-existing GOTO binary, annotate the invariant clause into 
+the GOTO binary, and produce the annotated binary as output.
+
+Loop invariant synthesized by the synthesizer must be inductive 
+in the sense that 
+1. the instrumented loop invariant checks success, no matter other 
+assertions fail or success;
+2. the base-case is from the whole GOTO program with a main 
+function as the entry point.
+
```

Comment:
Could you provide an example here?

Reply:
The first point I want to make here is that we only care about if the invariant is inductive, but not other checks (even if there is `assert(0);` in the loop). The purpose of the second point is to distinguish 1. the invariants are inductive for any inputs of the function, and 2. the invariants are inductive for only paths in the given GOTO binary.

As an example, ```k == 1``` is an inductive invariant in the following program. However, it is not always inductive if we only consider the function ```foo```---it doesn't hold when the parameter ```k``` is not 1.
```c
void foo(int k)
{
  for(int i = 0; i < 5; i++)
  {
    if(k != 1)
      k = 2;
  }
}

void main()
{
  foo(1);
}
```

You are right that the semantics of loop invariants is enough to illustrate the point. I will update the text and add a link to the semantics of loop invariants.

Reply:
I still few this section needs an example. Take a small program, check the program with pointer checks enable using a bound, synthesize the invariant, check the problem again without the bound. One of the most important part of the documentation is to walk potential users and future developers in the process of how to use this tool.

Reply:
I like the idea of examples.  It would also be good to discuss what the initial state is as part of defining / saying they are inductive.

Reply:
If the invariants are inferred in the context of a proof harness that does not fully cover the input space of the function that contains the loop, is there a risk that we could assume the invariant in a different context where it could actually not hold without realizing it?

Reply:
It could happen but it should not be silent as the check for whether it is a loop invariant should fail in the new context.  I presume we have tests for "`__CPROVER_invariant` that aren't actually true?